src/aarch64/kernel_machine.h

```diff
@@ -0,0 +1,375 @@
+#define ESR_ISS_ID_ABRT_FSC_SYNC_PARITY_ECC_L3 0x1f
+#define ESR_ISS_ID_ABRT_FSC_TLB_CONFLICT_ABORT 0x30
+
+#define ID_AA64ISAR0_EL1_RNDR_BITS        4
+#define ID_AA64ISAR0_EL1_RNDR_SHIFT       60
+#define ID_AA64ISAR0_EL1_RNDR_IMPLEMENTED 1 /* RNDR, RNDRRS MSRs */
+
+#define SCTLR_EL1_UCI     U64_FROM_BIT(26) /* trap cache instructions in EL0 */
+#define SCTLR_EL1_EE      U64_FROM_BIT(25) /* endianness for EL1 data / pt table */
+#define SCTLR_EL1_E0E     U64_FROM_BIT(24) /* endianness for EL0 data */
+#define SCTLR_EL1_WXN     U64_FROM_BIT(19) /* write implies execute never */
+#define SCTLR_EL1_I       U64_FROM_BIT(12) /* instruction cacheability (no effect) */
+#define SCTLR_EL1_SA0     U64_FROM_BIT(4) /* SP alignment fault enable for EL0 */
+#define SCTLR_EL1_SA      U64_FROM_BIT(3) /* SP alignment fault enable */
+#define SCTLR_EL1_C       U64_FROM_BIT(2) /* cacheability (no effect) */
+#define SCTLR_EL1_A       U64_FROM_BIT(1) /* alignment fault check enable */
+#define SCTLR_EL1_M       U64_FROM_BIT(0) /* MMU enable */
+
+#define SPSR_I U64_FROM_BIT(7)
+
+/* interrupt control */
+static inline void enable_interrupts(void)
+{
+    asm volatile("msr daifclr, #2");
+}
+
+static inline void disable_interrupts(void)
+{
+    asm volatile("msr daifset, #2");
+}
+
+static inline u64 irq_disable_save(void)
+{
+    register u32 daif;
+    asm volatile("mrs %0, daif; msr daifset, #2" : "=r"(daif));
+    return daif;
+}
+
+static inline void irq_restore(u64 flags)
+{
+    asm volatile("msr daif, %0" :: "r"(flags));
+}
+
+static inline void wait_for_interrupt(void)
+{
+    disable_interrupts();
+    asm volatile("dsb sy; wfi" ::: "memory");
+    enable_interrupts();
+}
+
+/* locking constructs */
+#include <lock.h>
+
+/* special register access */
+#define read_psr(reg) ({ register u64 r; asm volatile("mrs %0, " #reg : "=r"(r)); r;})
+#define write_psr(reg, v) do { asm volatile("msr " #reg ", %0" : : "r"(v)); } while (0)
+#define read_psr_s(rstr) ({ register u64 r; asm volatile("mrs %0, " rstr : "=r"(r)); r;})
+#define write_psr_s(rstr, v) do { asm volatile("msr " rstr ", %0" : : "r"(v)); } while (0)
+
+/* per-cpu info, saved contexts and stacks */
+typedef u64 *context;
+
+#define KERNEL_STACK_WORDS (KERNEL_STACK_SIZE / sizeof(u64))
+typedef struct kernel_context {
+    u64 stackbase[KERNEL_STACK_WORDS];
+    u64 frame[0];
+} *kernel_context;
+
+typedef struct cpuinfo {
+    /*** Fields accessed by low-level entry points. ***/
+    /* Don't move these without updating x18-relative accesses in crt0.s ***/
+
+    /* This points to the frame of the current, running context. +0 */
+    context running_frame;
+
+    /* Default frame and stack installed at kernel entry points (init,
+       syscall) and calls to runloop. +8 */
+    kernel_context kernel_context;
+
+    /* One temporary for syscall enter to use so that we don't need to touch the user stack. +16 */
+    u64 tmp;
+
+    /*** End of fields touched by kernel entries ***/
+
+    u32 id;
```

Comment:
Does it make sense to move the rest or some of the rest of this struct to an arch independent struct to remove redundancy and make a distinction between arch and arch-independent fields? It might be a lot of churn so perhaps not worth it.

Reply:
I agree that there should be a common struct, so I moved struct cpuinfo to kernel.h and made a 'struct cpuinfo_machine', which is defined in src/\<arch\>/kernel_machine.h, as its first field.